File: app/src/main/java/wangdaye/com/geometricweather/common/utils/DisplayUtils.java
package wangdaye.com.geometricweather.common.utils;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.media.ThumbnailUtils;
import android.os.Build;
import android.text.format.DateFormat;
import android.view.View;
import android.view.Window;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.Interpolator;
import android.view.animation.OvershootInterpolator;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Px;
import androidx.annotation.RequiresApi;
import androidx.annotation.Size;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.ColorUtils;

import java.util.Calendar;
import java.util.TimeZone;

import wangdaye.com.geometricweather.R;

/**
 * Display utils.
 * */

public class DisplayUtils {

    public static final Interpolator FLOATING_DECELERATE_INTERPOLATOR
            = new DecelerateInterpolator(1f);

    private static final int MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_PHONE = 512;
    private static final int MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_TABLET = 600;

    public static float dpToPx(Context context, float dp) {
        return dp * (context.getResources().getDisplayMetrics().densityDpi / 160f);
    }

    public static float spToPx(Context context, int sp) {
        return sp * (context.getResources().getDisplayMetrics().scaledDensity);
    }

    public static void setSystemBarStyle(Context context, Window window,
                                         boolean statusShader, boolean lightStatus,
                                         boolean navigationShader, boolean lightNavigation) {
        setSystemBarStyle(context, window,
                false, statusShader, lightStatus, navigationShader, lightNavigation);
    }

    public static void setSystemBarStyle(Context context, Window window, boolean miniAlpha,
                                         boolean statusShader, boolean lightStatus,
                                         boolean navigationShader, boolean lightNavigation) {

        // statusShader &= Build.VERSION.SDK_INT < Build.VERSION_CODES.Q;
        lightStatus &= Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
        navigationShader &= Build.VERSION.SDK_INT < Build.VERSION_CODES.Q;
        lightNavigation &= Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;

        int visibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_STABLE;
        if (lightStatus) {
            visibility |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
        }
        if (lightNavigation) {
            visibility |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
        }
        window.getDecorView().setSystemUiVisibility(visibility);

        setSystemBarColor(context, window, miniAlpha, statusShader, lightStatus, navigationShader, lightNavigation);
    }

    public static void setSystemBarColor(Context context, Window window, boolean miniAlpha,
                                         boolean statusShader, boolean lightStatus,
                                         boolean navigationShader, boolean lightNavigation) {

        // statusShader &= Build.VERSION.SDK_INT < Build.VERSION_CODES.Q;
        lightStatus &= Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
        navigationShader &= Build.VERSION.SDK_INT < Build.VERSION_CODES.Q;
        lightNavigation &= Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;

        if (!statusShader) {
            window.setStatusBarColor(Color.argb(0, 0, 0, 0));
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            window.setStatusBarColor(getStatusBarColor23(context, lightStatus, miniAlpha));
        } else {
            window.setStatusBarColor(getStatusBarColor21());
        }
        if (!navigationShader) {
            window.setNavigationBarColor(Color.argb(0, 0, 0, 0));
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            window.setNavigationBarColor(getStatusBarColor26(context, lightNavigation, miniAlpha));
        } else {
            window.setNavigationBarColor(getNavigationBarColor21());
        }
    }

    @ColorInt
    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private static int getStatusBarColor21() {
        return ColorUtils.setAlphaComponent(Color.BLACK, (int) (0.1 * 255));
    }

    @ColorInt
    @RequiresApi(Build.VERSION_CODES.M)
    private static int getStatusBarColor23(Context context, boolean light, boolean miniAlpha) {
        if (miniAlpha) {
            return light
                    ? ColorUtils.setAlphaComponent(Color.WHITE, (int) (0.2 * 255))
                    : ColorUtils.setAlphaComponent(Color.BLACK, (int) (0.1 * 255));
        }
        return ColorUtils.setAlphaComponent(
                ContextCompat.getColor(context, light ? R.color.colorRoot_light : R.color.colorRoot_dark),
                (int) (0.8 * 255)
        );
    }

    @ColorInt
    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private static int getNavigationBarColor21() {
        return ColorUtils.setAlphaComponent(Color.BLACK, (int) (0.1 * 255));
    }

    @ColorInt
    @RequiresApi(Build.VERSION_CODES.O)
    private static int getStatusBarColor26(Context context, boolean light, boolean miniAlpha) {
        if (miniAlpha) {
            return light
                    ? ColorUtils.setAlphaComponent(Color.WHITE, (int) (0.2 * 255))
                    : ColorUtils.setAlphaComponent(Color.BLACK, (int) (0.1 * 255));
        }
        return ColorUtils.setAlphaComponent(
                ContextCompat.getColor(context, light ? R.color.colorRoot_light : R.color.colorRoot_dark),
                (int) (0.8 * 255)
        );
    }

    public static boolean isTabletDevice(Context context) {
        return (context.getResources().getConfiguration().screenLayout
                & Configuration.SCREENLAYOUT_SIZE_MASK) >= Configuration.SCREENLAYOUT_SIZE_LARGE;
    }

    public static boolean isLandscape(Context context) {
        return context.getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
    }

    public static boolean isRtl(Context context) {
        return context.getResources().getConfiguration().getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
    }

    public static boolean isDarkMode(Context context) {
        return (context.getResources().getConfiguration().uiMode
                & Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES;
    }

    @NonNull
    public static Bitmap drawableToBitmap(@NonNull Drawable drawable) {
        Bitmap bitmap = Bitmap.createBitmap(
                drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(),
                Bitmap.Config.ARGB_8888
        );

        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    @ColorInt
    public static int bitmapToColorInt(@NonNull Bitmap bitmap) {
        return ThumbnailUtils.extractThumbnail(bitmap, 1, 1)
                .getPixel(0, 0);
    }

    public static boolean isLightColor(@ColorInt int color) {
        int alpha = 0xFF << 24;
        int grey = color;
        int red = ((grey & 0x00FF0000) >> 16);
        int green = ((grey & 0x0000FF00) >> 8);
        int blue = (grey & 0x000000FF);

        grey = (int) (red * 0.3 + green * 0.59 + blue * 0.11);
        grey = alpha | (grey << 16) | (grey << 8) | grey;
        return grey > 0xffbdbdbd;
    }

    @Px
    public static int getTabletListAdaptiveWidth(Context context, @Px int width) {
        if (!isTabletDevice(context) && !isLandscape(context)) {
            return width;
        }
        return (int) Math.min(
                width,
                DisplayUtils.dpToPx(
                        context,
                        isTabletDevice(context)
                                ? MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_TABLET
                                : MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_PHONE
                )
        );
    }

    @ColorInt
    public static int blendColor(@ColorInt int foreground, @ColorInt int background) {
        int scr = Color.red(foreground);
        int scg = Color.green(foreground);
        int scb = Color.blue(foreground);
        int sa = foreground >>> 24;
        int dcr = Color.red(background);
        int dcg = Color.green(background);
        int dcb = Color.blue(background);
        int color_r = dcr * (0xff - sa) / 0xff + scr * sa / 0xff;
        int color_g = dcg * (0xff - sa) / 0xff + scg * sa / 0xff;
        int color_b = dcb * (0xff - sa) / 0xff + scb * sa / 0xff;
        return ((color_r << 16) + (color_g << 8) + color_b) | (0xff000000);
    }

    public static boolean is12Hour(Context context) {
        return !DateFormat.is24HourFormat(context);
    }

    public static boolean isDaylight(TimeZone timeZone) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeZone(timeZone);
        int time = 60 * calendar.get(Calendar.HOUR_OF_DAY) + calendar.get(Calendar.MINUTE);

        int sr = 60 * 6;
        int ss = 60 * 18;
        return sr < time && time < ss;
    }

    // translationY, scaleX, scaleY
    @Size(3)
    public static Animator[] getFloatingOvershotEnterAnimators(View view) {
        return getFloatingOvershotEnterAnimators(view, 1.5f);
    }

    @Size(3)
    public static Animator[] getFloatingOvershotEnterAnimators(View view, float overshootFactor) {
        return getFloatingOvershotEnterAnimators(view, overshootFactor,
                view.getTranslationY(), view.getScaleX(), view.getScaleY());
    }

    @Size(3)
    public static Animator[] getFloatingOvershotEnterAnimators(View view,
                                                               float overshootFactor,
                                                               float translationYFrom,
                                                               float scaleXFrom,
                                                               float scaleYFrom) {
        Animator translation = ObjectAnimator.ofFloat(
                view, "translationY", translationYFrom, 0f);
        translation.setInterpolator(new OvershootInterpolator(overshootFactor));

        Animator scaleX = ObjectAnimator.ofFloat(
                view, "scaleX", scaleXFrom, 1f);
        scaleX.setInterpolator(FLOATING_DECELERATE_INTERPOLATOR);

        Animator scaleY = ObjectAnimator.ofFloat(
                view, "scaleY", scaleYFrom, 1f);
        scaleY.setInterpolator(FLOATING_DECELERATE_INTERPOLATOR);

        return new Animator[] {translation, scaleX, scaleY};
    }

    public static void getVisibleDisplayFrame(View view, Rect rect) {
        /*
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            DisplayMetrics metrics = new DisplayMetrics();
            WindowManager wm = (WindowManager) view.getContext().getSystemService(
                    Context.WINDOW_SERVICE);
            wm.getDefaultDisplay().getRealMetrics(metrics);

            WindowInsets insets = view.getRootWindowInsets();

            rect.set(
                    insets.getSystemWindowInsetLeft(),
                    insets.getSystemWindowInsetTop(),
                    metrics.widthPixels - insets.getSystemWindowInsetRight(),
                    metrics.heightPixels - insets.getSystemWindowInsetBottom()
            );
        } else {
            view.getWindowVisibleDisplayFrame(rect);
        }
        */
        // looks like has a good performance.
        view.getWindowVisibleDisplayFrame(rect);
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/utils/DisplayUtils.kt
package wangdaye.com.geometricweather.common.utils

import android.animation.Animator
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.media.ThumbnailUtils
import android.os.Build
import android.text.format.DateFormat
import android.view.View
import android.view.Window
import android.view.animation.DecelerateInterpolator
import android.view.animation.Interpolator
import android.view.animation.OvershootInterpolator
import androidx.annotation.ColorInt
import androidx.annotation.Px
import androidx.annotation.Size
import androidx.annotation.StyleRes
import androidx.core.graphics.ColorUtils
import androidx.core.view.WindowCompat
import com.google.android.material.resources.TextAppearance
import kotlin.math.min

object DisplayUtils {

    @JvmField
    val FLOATING_DECELERATE_INTERPOLATOR: Interpolator = DecelerateInterpolator(1f)

    private const val MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_PHONE = 512
    private const val MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_TABLET = 600

    const val DEFAULT_CARD_LIST_ITEM_ELEVATION_DP = 2f

    @JvmStatic
    fun dpToPx(context: Context, dp: Float): Float {
        return dp * (context.resources.displayMetrics.densityDpi / 160f)
    }

    @JvmStatic
    fun spToPx(context: Context, sp: Int): Float {
        return sp * context.resources.displayMetrics.scaledDensity
    }

    @JvmStatic
    fun pxToDp(context: Context, @Px px: Int): Float {
        return px / (context.resources.displayMetrics.densityDpi / 160f)
    }

    @JvmStatic
    fun setSystemBarStyle(
        context: Context,
        window: Window,
        lightStatus: Boolean,
        lightNavigation: Boolean
    ) {
        setSystemBarStyle(
            context,
            window,
            false,
            lightStatus,
            false,
            lightNavigation
        )
    }

    @JvmStatic
    fun setSystemBarStyle(
        context: Context,
        window: Window,
        statusShader: Boolean,
        lightStatus: Boolean,
        navigationShader: Boolean,
        lightNavigation: Boolean
    ) {
        var statusLight = lightStatus
        var statusShade = statusShader
        var navLight = lightNavigation
        var navShade = navigationShader

        WindowCompat.setDecorFitsSystemWindows(window, false)

        val insetsController = WindowCompat.getInsetsController(window, window.decorView)
        insetsController.isAppearanceLightStatusBars = statusLight
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            insetsController.isAppearanceLightNavigationBars = navLight
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            statusLight = false
            statusShade = true
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            navLight = false
            navShade = true
        }
        navShade = navShade && Build.VERSION.SDK_INT < Build.VERSION_CODES.Q

        if (!statusShade) {
            window.statusBarColor = Color.TRANSPARENT
        } else {
            window.statusBarColor = ColorUtils.setAlphaComponent(
                if (statusLight) Color.WHITE else Color.BLACK,
                ((if (statusLight) 0.5 else 0.2) * 255).toInt()
            )
        }
        if (!navShade) {
            window.navigationBarColor = Color.TRANSPARENT
        } else {
            window.navigationBarColor = ColorUtils.setAlphaComponent(
                if (navLight) Color.WHITE else Color.BLACK,
                ((if (navLight) 0.5 else 0.2) * 255).toInt()
            )
        }
    }

    @JvmStatic
    fun isTabletDevice(context: Context): Boolean {
        return (context.resources.configuration.screenLayout
            and Configuration.SCREENLAYOUT_SIZE_MASK) >= Configuration.SCREENLAYOUT_SIZE_LARGE
    }

    @JvmStatic
    fun isLandscape(context: Context): Boolean {
        return context.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
    }

    @JvmStatic
    fun isRtl(context: Context): Boolean {
        return context.resources.configuration.layoutDirection == View.LAYOUT_DIRECTION_RTL
    }

    @JvmStatic
    fun isDarkMode(context: Context): Boolean {
        return (context.resources.configuration.uiMode
            and Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES
    }

    @JvmStatic
    fun drawableToBitmap(drawable: Drawable): Bitmap {
        val bitmap = Bitmap.createBitmap(
            drawable.intrinsicWidth,
            drawable.intrinsicHeight,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
        drawable.draw(canvas)
        return bitmap
    }

    @JvmStatic
    @ColorInt
    fun bitmapToColorInt(bitmap: Bitmap): Int {
        return ThumbnailUtils.extractThumbnail(bitmap, 1, 1).getPixel(0, 0)
    }

    @JvmStatic
    fun isLightColor(@ColorInt color: Int): Boolean {
        val alpha = 0xFF shl 24
        var grey = color
        val red = (grey and 0x00FF0000) shr 16
        val green = (grey and 0x0000FF00) shr 8
        val blue = grey and 0x000000FF
        grey = (red * 0.3 + green * 0.59 + blue * 0.11).toInt()
        grey = alpha or (grey shl 16) or (grey shl 8) or grey
        return grey > 0xffbdbdbd.toInt()
    }

    @JvmStatic
    @Px
    fun getTabletListAdaptiveWidth(context: Context, @Px width: Int): Int {
        if (!isTabletDevice(context) && !isLandscape(context)) {
            return width
        }
        return min(
            width.toDouble(),
            dpToPx(
                context,
                (
                    if (isTabletDevice(context)) MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_TABLET
                    else MAX_TABLET_ADAPTIVE_LIST_WIDTH_DIP_PHONE
                    ).toFloat()
            ).toDouble()
        ).toInt()
    }

    @JvmStatic
    @ColorInt
    fun blendColor(@ColorInt foreground: Int, @ColorInt background: Int): Int {
        val scr = Color.red(foreground)
        val scg = Color.green(foreground)
        val scb = Color.blue(foreground)
        val sa = foreground ushr 24
        val dcr = Color.red(background)
        val dcg = Color.green(background)
        val dcb = Color.blue(background)
        val colorR = dcr * (0xff - sa) / 0xff + scr * sa / 0xff
        val colorG = dcg * (0xff - sa) / 0xff + scg * sa / 0xff
        val colorB = dcb * (0xff - sa) / 0xff + scb * sa / 0xff
        return (colorR shl 16) + (colorG shl 8) + colorB or -0x1000000
    }

    @JvmStatic
    fun is12Hour(context: Context): Boolean {
        return !DateFormat.is24HourFormat(context)
    }

    @JvmStatic
    @Size(3)
    fun getFloatingOvershotEnterAnimators(view: View): Array<Animator> {
        return getFloatingOvershotEnterAnimators(view, 1.5f)
    }

    @JvmStatic
    @Size(3)
    fun getFloatingOvershotEnterAnimators(view: View, overshootFactor: Float): Array<Animator> {
        return getFloatingOvershotEnterAnimators(
            view,
            overshootFactor,
            view.translationY,
            view.scaleX,
            view.scaleY
        )
    }

    @JvmStatic
    @Size(3)
    fun getFloatingOvershotEnterAnimators(
        view: View,
        overshootFactor: Float,
        translationYFrom: Float,
        scaleXFrom: Float,
        scaleYFrom: Float
    ): Array<Animator> {
        val translation = ObjectAnimator.ofFloat(view, "translationY", translationYFrom, 0f)
        translation.interpolator = OvershootInterpolator(overshootFactor)
        val scaleX = ObjectAnimator.ofFloat(view, "scaleX", scaleXFrom, 1f)
        scaleX.interpolator = FLOATING_DECELERATE_INTERPOLATOR
        val scaleY = ObjectAnimator.ofFloat(view, "scaleY", scaleYFrom, 1f)
        scaleY.interpolator = FLOATING_DECELERATE_INTERPOLATOR
        return arrayOf(translation, scaleX, scaleY)
    }

    @JvmStatic
    fun getVisibleDisplayFrame(view: View, rect: Rect) {
        view.getWindowVisibleDisplayFrame(rect)
    }

    @SuppressLint("RestrictedApi", "VisibleForTests")
    @JvmStatic
    fun getTypefaceFromTextAppearance(context: Context, @StyleRes textAppearanceId: Int): Typeface {
        return TextAppearance(context, textAppearanceId).getFont(context)
    }

    @JvmStatic
    @ColorInt
    fun getWidgetSurfaceColor(
        elevationDp: Float,
        @ColorInt tintColor: Int,
        @ColorInt surfaceColor: Int
    ): Int {
        if (elevationDp == 0f) {
            return surfaceColor
        }
        val foreground = ColorUtils.setAlphaComponent(
            tintColor,
            (((4.5f * Math.log((elevationDp + 1).toDouble())) + 2f) / 100f * 255).toInt()
        )
        return blendColor(foreground, surfaceColor)
    }
}
